snapshot/pkg/volume/openebs/processor.go

```diff
@@ -0,0 +1,312 @@
+
+	// TODO implement snapshot-info based on snapshotID
+	resp, err := h.SnapshotInfo(snapshotData.Spec.PersistentVolumeRef.Name, snapshotID)
+
+	if err != nil {
+		glog.Errorf("failed to describe snapshot:%v", snapshotID)
+	}
+
+	glog.V(1).Infof("snapshot details:%v", string(resp))
+
+	if len(snapshotData.Status.Conditions) == 0 {
+		return nil, false, fmt.Errorf("No status condtions in VoluemSnapshotData for openebs snapshot type")
+	}
+
+	lastCondIdx := len(snapshotData.Status.Conditions) - 1
+	retCondType := crdv1.VolumeSnapshotConditionError
+
+	switch snapshotData.Status.Conditions[lastCondIdx].Type {
+	case crdv1.VolumeSnapshotDataConditionReady:
+		retCondType = crdv1.VolumeSnapshotConditionReady
+	case crdv1.VolumeSnapshotDataConditionPending:
+		retCondType = crdv1.VolumeSnapshotConditionPending
+		// Error out.
+	}
+	retCond := []crdv1.VolumeSnapshotCondition{
+		{
+			Status:             snapshotData.Status.Conditions[lastCondIdx].Status,
+			Message:            snapshotData.Status.Conditions[lastCondIdx].Message,
+			LastTransitionTime: snapshotData.Status.Conditions[lastCondIdx].LastTransitionTime,
+			Type:               retCondType,
+		},
+	}
+	return &retCond, true, nil
+}
+
+// FindSnapshot finds a VolumeSnapshot by matching metadata
+func (h *openEBSPlugin) FindSnapshot(tags *map[string]string) (*crdv1.VolumeSnapshotDataSource, *[]crdv1.VolumeSnapshotCondition, error) {
+	glog.Infof("FindSnapshot by tags: %#v", *tags)
+
+	// TODO: Implement FindSnapshot
+	return &crdv1.VolumeSnapshotDataSource{
+		OpenEBSSnapshot: &crdv1.OpenEBSVolumeSnapshotSource{
+			SnapshotID: "",
+		},
+	}, nil, nil
+}
+
+// SnapshotRestore restore to any created snapshot
+func (h *openEBSPlugin) SnapshotRestore(snapshotData *crdv1.VolumeSnapshotData,
+	pvc *v1.PersistentVolumeClaim,
+	pvName string,
+	parameters map[string]string,
+) (*v1.PersistentVolumeSource, map[string]string, error) {
+
+	// GetMayaService get the maya-service endpoint
+	//err := GetMayaService()
+	//if err != nil {
+	//	return nil, nil, err
+	//}
+	if snapshotData == nil || snapshotData.Spec.OpenEBSSnapshot == nil {
+		return nil, nil, fmt.Errorf("Invalid Snapshot spec")
+	}
+	if pvc == nil {
+		return nil, nil, fmt.Errorf("Invalid PVC spec")
+	}
+
+	// restore snapshot to a PV
+	snapshotID := snapshotData.Spec.OpenEBSSnapshot.SnapshotID
+	pvRefName := snapshotData.Spec.PersistentVolumeRef.Name
+	pvRefNamespace := snapshotData.Spec.PersistentVolumeRef.Namespace
+
+	var oldvolume, newvolume mayav1.Volume
+	var openebsVol mApiv1.OpenEBSVolume
+	volumeSpec := mayav1.VolumeSpec{}
+
+	volSize := pvc.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	volumeSpec.Metadata.Labels.Storage = volSize.String()
+	volumeSpec.Metadata.Labels.StorageClass = *pvc.Spec.StorageClassName
+	volumeSpec.Metadata.Labels.Namespace = pvc.Namespace
+	volumeSpec.Metadata.Name = pvName
+
+	err := openebsVol.ListVolume(pvRefName, pvRefNamespace, &oldvolume)
```

Comment:
Snapshot Provisioner should pass the source volume name to the maya-apiserver. The API server should have the intelligence to retrieve the required clone IP. This way the snapshot provisioner doesn't have to depend on the internal details of the volume types supported by the maya-apiserver.

Reply:
The storage class name passed should be that of the source volume, so that the maya-apsierver can bring up the cloned volume with the same parameters as source volume. In case the cloned volume needs to override some values, they should be mentioned in the PVC.

Reply:
Added , now snapshot-provisioner  will fetch the SC from the source volume.